File: src/main/java/io/github/alexiscomete/lapinousecond/roles_update/RolesEnum.java
package io.github.alexiscomete.lapinousecond.roles_update;

import io.github.alexiscomete.lapinousecond.entity.Player;
import org.javacord.api.entity.permission.Role;
import org.javacord.api.entity.server.Server;
import org.javacord.api.entity.user.User;

import java.util.ArrayList;
import java.util.List;

public enum RolesEnum {

    ADMIN("admin", "Administrateur du serveur Discord", new String[]{"admin", "orgna", "orga"}, 80, /* 5 heures */ 43_200), // pas besoin de mettre le nom complet du role car le contain ignore les espaces
    MODO("modo", "Modérateur du serveur Discord", new String[]{"modo", "modé"}, 60, 43_200),
    MEMBER("member", "Membre du serveur Discord", new String[]{"memb"}, 10, /* 1h */3_600),
    PARTICIPANT("participant", "Participant du serveur Discord", new String[]{"part", "commu"}, 50, /* 3 heures */ 10_800),
    VISITOR("visitor", "Visiteur du serveur Discord", new String[]{"visit"}, 5, /* 1h */3_600),
    CITOYEN("citoyen", "Citoyen du Dibistan / de la région / du département", new String[]{"citoy"}, 10, /* 3 heures */ 10_800),
    AMBASSADOR("ambassadeur", "Ambassadeur", new String[]{"ambassad"}, 10, /* 3 heures */ 10_800),
    DELEGATE("delegate", "Représentant", new String[]{"delag", "repr"}, 10, /* 3 heures */ 10_800);

    public final String name;
    public final String description;
    // les alias
    public final String[] aliases;
    public final int salary;
    public final int coolDownSize; // en secondes

    RolesEnum(String name, String description, String[] aliases, int salary, int coolDownSize) {
        this.name = name;
        this.description = description;
        this.aliases = aliases;
        this.salary = salary;
        this.coolDownSize = coolDownSize;
    }

    public static boolean check(String[] strings, ArrayList<RolesEnum> roles, long serverID) {
        for (String string : strings) {
            boolean perm;
            switch (string) {
                case "MANAGE_ADMIN":
                    perm = false;
                    for (RolesEnum role : roles) {
                        if (role.getServerID() == serverID && role.getProgName().equals("SERVER_OWNER")) {
                            perm = true;
                            break;
                        }
                    }
                    if (!perm) {
                        return false;
                    }
                    break;
                case "MANAGE_ROLES_SERVER":
                    perm = false;
                    for (RolesEnum role : roles) {
                        if (role.getServerID() == serverID && (role.getProgName().equals("SERVER_OWNER") || role.getProgName().equals("PROJECT_ADMIN") || role.getProgName().equals("SERVER_ADMIN"))) {
                            perm = true;
                            break;
                        }
                    }
                    if (!perm) {
                        return false;
                    }
                    break;
            }
        }
        return true;
    }

    public boolean check(User user, Server server) {
        List<Role> roles = user.getRoles(server);
        for (String alias : aliases) {
            for (Role role : roles) {
                if (role.getName().toLowerCase().contains(alias)) {
                    return true;
                }
            }
        }
        return false;
    }
}

File: src/main/java/io/github/alexiscomete/lapinousecond/entity/roles/RolesEnum.kt
package io.github.alexiscomete.lapinousecond.entity.roles

import org.javacord.api.entity.server.Server
import org.javacord.api.entity.user.User
import java.util.*

enum class RolesEnum(
    val displayName: String, val description: String, // les alias
    private val aliases: Array<String>, val salary: Int, // en secondes
    val coolDownSize: Int
) {
    ADMIN(
        "Admin",
        "Administrateur du serveur Discord",
        arrayOf("admin", "orgna", "orga"),
        80,  /* 5 heures */
        43200
    ),  // pas besoin de mettre le nom complet du role car le contain ignore les espaces
    MODO(
        "Modérateur",
        "Modérateur du serveur Discord (ou un chanceux qui a un rôle qui bug)",
        arrayOf("modo", "modé"),
        60,
        43200
    ),
    MEMBER(
        "Membre",
        "Membre du serveur Discord",
        arrayOf("memb"),
        10,  /* 1h */
        3600
    ),
    PARTICIPANT(
        "Participant",
        "Participant du serveur Discord",
        arrayOf("part", "commu"),
        50,  /* 3 heures */
        10800
    ),
    VISITOR(
        "Visiteur",
        "Visiteur du serveur Discord, de passage ici puis ailleurs, et enfin là-bas",
        arrayOf("visit"),
        5,  /* 1h */
        3600
    ),
    CITOYEN(
        "Citoyen",
        "Citoyen de la ville du serveur",
        arrayOf("citoy"),
        10,  /* 3 heures */
        10800
    ),
    AMBASSADOR(
        "Ambassadeur",
        "Ambassadeur d'un autre serveur",
        arrayOf("ambassad"),
        10,  /* 3 heures */
        10800
    ),
    DELEGATE(
        "Délégué",
        "Représentant",
        arrayOf("delag", "repr"),
        10,  /* 3 heures */
        10800
    );

    fun check(user: User, server: Server?): Boolean {
        val roles = user.getRoles(server)
        for (alias in aliases) {
            for (role in roles) {
                if (role.name.lowercase(Locale.getDefault()).contains(alias)) {
                    return true
                }
            }
        }
        return false
    }

    companion object {
        fun check(strings: Array<String>, user: User, server: Server): Boolean {
            for (string in strings) {
                return when (string) {
                    "MANAGE_ADMIN" -> server.owner.get() == user
                    "MANAGE_ROLES_SERVER" -> ADMIN.check(user, server)
                    else -> {
                        println("WARNING : $string n'est pas une permission")
                        true
                    }
                }
            }
            return true
        }

        fun getRoles(user: User, server: Server?): ArrayList<RolesEnum> {
            val roles = ArrayList<RolesEnum>()
            for (role in entries) {
                if (role.check(user, server)) {
                    roles.add(role)
                }
            }
            return roles
        }
    }
}
